### ISSUE TITLE
Allow local xref definitions in respec config

### ISSUE THREAD

**Issue by tidoust** [w3c/respec#3227]
Inevitably, specs need to reference terms defined in specifications that are not part of the xref database, be it only temporarily while the spec gets added to the xref database. To do that, several specs resort to having a Terminology section. For instance, see the first few terms in the [Terminology section of the Presentation API](https://w3c.github.io/presentation-api/#terminology) (defined in ECMAScript and in several RFC).

The net result of such a construct is that these terms appear in the automatically generated index in the list of "Terms defined by this specification". Ideally, they would rather appear in the list of "Terms defined by reference" and there would be no need to create a terminology section in the first place.

For that to be possible, there needs to be a way to describe "local xref definitions" somehow that could be used to link things up and generate the right index, a bit like it's possible to specify a local biblio. That is doable in Bikeshed (see [Providing Custom Definitions](https://tabatkins.github.io/bikeshed/#custom-dfns)), but not (yet :)) in ReSpec.

**Comment by sidvishnoi** [comment]
We've given this thoughts earlier while designing the xref feature and decided not to support `localXref` or similar, as we will be able to direct more specs to get into the database this way. But I agree, it's inevitable that some will not get into the database.

Instead, we decided to make use of the existing `data-cite` attribute. In short:
```html
Instead of <dfn>foo</dfn>,
Try: <dfn data-cite="SPEC#frag">foo</dfn>
```

For a more complete example, see: [gist](https://gist.github.com/sidvishnoi/a7a406861be71cf7f6873d78c7e3e8c6), [live demo](https://gistcdn.githack.com/sidvishnoi/a7a406861be71cf7f6873d78c7e3e8c6/raw/6eb94c1c19d9c87e41e6c9faf2ad899e30328120/index.html)

These definitions can be inline, so it's not required to create a terminology section. We need to define them once only and they can be referenced using the regular `<a>term</a>` or `[= term =]` (no need to repeat `data-cite`). `data-dfn-for` and `data-dfn-type` also work fine.

(ReSpec docs are still missing things, I agree.)

**Comment by tidoust** [comment]
Thanks. That's almost perfect! It does not necessarily require a terminology section, but it still requires the presence of an arbitrary `<dfn>` and that dfn then appears in bold for no good reason (it is not a real definition and nothing links to it).

I wonder whether one could make do without the dfn, as in:

```html
Only once: <a data-cite="SPEC#frag">foo</a>
Other occurrences: [= foo =]
```

Authors would still need to single out one specific occurrence in the text, but that occurrence would still be rendered as a regular link.

The above example does not work because the code cannot resolve `[= foo =]` for now, but I note that the generated index seems correct, so it may just be a matter of considering `<a>` elements with `data-cite` attributes when resolving links. I do not know whether that could trigger side effects though...

An alternative could be to continue to require the `<dfn>`, but to drop it in the generated markup so that it ends up being rendered as a regular link.

PS: Well, and it could be good if the documentation didn't essentially say "don't use data-cite" in bold ;) Something like "use xref whenever possible. If that is really not possible, here is what you can do..."

**Comment by sidvishnoi** [comment]
> I wonder whether one could make do without the dfn, as in:

It's definitely better, but requires significant coding and testing effort (potential breaking change) :disappointed: 

> An alternative could be to continue to require the <dfn>, but to drop it in the generated markup so that it ends up being rendered as a regular link.

I prefer this approach. Still need to check it for breaking changes though (dfn panels might depend on it). 
As a temporary fix, it can be done in `postProcess` (if you try it, let us know here).

---

> if the documentation didn't essentially say "don't use data-cite" in bold ;)

:rofl:  Point taken. 

FWIW, there is a note at the bottom of page in common errors section:
> Note: Terms from ECMA/IETF specifications are not presently available in the terms database. Use the `data-cite` attribute to reference those terms.

**Comment by tidoust** [comment]
> As a temporary fix, it can be done in postProcess (if you try it, let us know here).

I haven't integrated it in any real spec yet, so there may be corner cases that need to be considered, but here is a `postProcess` function that seems to do the job:

```js
function replaceDataCiteDfns() {
  [...document.querySelectorAll('dfn[data-cite] > a')].forEach(el => {
    // Copy over "data-" attributes from the dfn
    // (step is not required but that seems cleaner)
    el.parentNode.getAttributeNames().forEach(name => {
      if (name.startsWith("data-")) {
        el.setAttribute(name, el.parentNode.getAttribute(name));
      }
    });

    // Drop the side panel that got created for the dfn, not to end up
    // with a broken link that would irk the linter
    const panel = document.getElementById(`dfn-panel-for-${el.parentNode.id}`);
    if (panel) {
      panel.parentNode.removeChild(panel);
    }

    // Replace the dfn element with the link itself
    el.parentNode.parentNode.replaceChild(el, el.parentNode);
  });
}
```

I note that the side dfn panel that ReSpec creates for `data-cite` dfns is basically useless: there is no way to show it (or is there?), and it seems to miss references to the term in any case...

To implement the `<a data-cite>` only approach, one can do the reverse operation in `preProcess`, e.g. with the following function:

```js
function createDataCiteDfns() {
  [...document.querySelectorAll('a[data-cite]')].forEach(el => {
    const link = document.createElement('dfn');
    el.getAttributeNames().forEach(name => {
      link.setAttribute(name, el.getAttribute(name));
    });
    let child = el.firstChild;
    while (child) {
      const nextChild = child.nextSibling;
      link.appendChild(child);
      child = nextChild;
    }
    el.parentNode.replaceChild(link, el);
  });
}
```

By the way, is there a way to couple `data-cite` with `data-dfn-type` to tell ReSpec that the cited term is an IDL term? Obviously, such terms should always be in the xref database, but the HTML spec does not yet export all these terms properly and there are remaining exceptions in practice. I can wrap the `data-cite` term definition in `<code>` so that ReSpec wraps references to it as well, but references still need to use `[=foo=]`, instead of `{{foo}}`.

**Comment by sidvishnoi** [comment]
Both temporary fixes look fine to me. I think it is something we will support in ReSpec (need to make time for it!).
As a nit, `panel.parentNode.removeChild(panel)` could be `panel.remove()`.

> I note that the side dfn panel that ReSpec creates for `data-cite` dfns is basically useless: there is no way to show it (or is there?), and it seems to miss references to the term in any case...

They should not be have created (filed https://github.com/w3c/respec/issues/3254). Though, dfn panels for `dfn[data-cite]` are shown in `<section id="index">`.

> By the way, is there a way to couple `data-cite` with `data-dfn-type` to tell ReSpec that the cited term is an IDL term? 

It should be working already.

```html
<p>
  <dfn data-dfn-type="interface" data-cite="ECMASCRIPT#current-realm">CurrentRealm</dfn> (not really an interface) in [[ECMASCRIPT]].
</p>
<p>[= CurrentRealm =] will not link.</p>
<p>{{ CurrentRealm }} will link to right external definition.</p>
```

**Comment by tidoust** [comment]
> Though, dfn panels for `dfn[data-cite]` are shown in `<section id="index">`.

Yes, these panels are correctly created and contain the right info AFAICT. The `postProcess` code does not remove them.

> > By the way, is there a way to couple `data-cite` with `data-dfn-type` to tell ReSpec that the cited term is an IDL term?
> 
> It should be working already.

Doh, I must have done something wrong when I tried it. I confirm that works just fine! 👍

**Comment by marcoscaceres** [comment]
Thanks @tidoust. As discussed, the `<dfn data-cite="SPEC#frag">term</dfn>` workaround covers the core use case. The remaining polish (dropping the dfn element in output) is minor. Closing as resolved by the existing workaround.
